pkg/yarn/client/ipc/kerberos.go

```diff
@@ -0,0 +1,285 @@
+	ClientUserName string
+	ClientRealm    string
+}
+
+func IsEnableKerberos() bool {
+	if v, ok := os.LookupEnv(EnvEnableKerberos); ok {
+		switch strings.ToLower(strings.TrimSpace(v)) {
+		case "1", "true", "yes", "y", "on", "TRUE":
+			return true
+		case "0", "false", "no", "n", "off", "FALSE":
+			return false
+		}
+	}
+	return false
+}
+
+func GetKerberosConf() KerberosConf {
+	conf := KerberosConf{
+		Krb5ConfPath:   DefaultKrb5ConfPath,
+		KeytabPath:     DefaultKeytabPath,
+		ClientUserName: DefaultClientUserName,
+		ClientRealm:    DefaultRealm,
+	}
+
+	if v, ok := os.LookupEnv(EnvKrb5ConfPath); ok && v != "" {
+		conf.Krb5ConfPath = v
+	}
+
+	if v, ok := os.LookupEnv(EnvKeytabPath); ok && v != "" {
+		conf.KeytabPath = v
+	}
+
+	if v, ok := os.LookupEnv(EnvClientUserName); ok && v != "" {
+		conf.ClientUserName = v
+	}
+
+	if v, ok := os.LookupEnv(EnvClientRealm); ok && v != "" {
+		conf.ClientRealm = v
+	}
+
+	return conf
+}
+
+func negotiateKerberosAuth(client *Client, con *connection) error {
+	// SASL NEGOTIATE
+	negState := hadoop_common.RpcSaslProto_NEGOTIATE
+	if err := sendSaslMessage(client, con, &hadoop_common.RpcSaslProto{State: &negState}); err != nil {
+		return err
+	}
+
+	resp, err := receiveSaslMessage(client, con)
+	if err != nil {
+		return err
+	}
+
+	var krbAuth *hadoop_common.RpcSaslProto_SaslAuth
+	for i, a := range resp.Auths {
+		klog.Infof("Server support auth [%d]: Method=%s, Mechanism=%s, Protocol=%s, ServerID=%s", i, a.GetMethod(), a.GetMechanism(), a.GetProtocol(), a.GetServerId())
+		if a.GetMethod() == "KERBEROS" && a.GetMechanism() == "GSSAPI" {
+			krbAuth = a
+		}
+	}
+	if krbAuth == nil {
+		klog.Errorf("kerberos with gssapi is not supported")
+		return fmt.Errorf("kerberos with gssapi is not supported")
+	}
+
+	// Kerberos Handshake
+	spn := fmt.Sprintf("%s/%s", krbAuth.GetProtocol(), krbAuth.GetServerId())
+	krb5Conf := GetKerberosConf()
+	token, sessionKey, err := generateGssapiKerberosToken(krb5Conf, spn)
+	if err != nil {
+		klog.Errorf("failed to generate gss-api kerberos token: %v", err)
+		return err
+	}
+
+	// SASL INITIATE
+	saslInitiateState := hadoop_common.RpcSaslProto_INITIATE
+	saslInitiateMessage := hadoop_common.RpcSaslProto{
+		State: &saslInitiateState,
+		Token: token,
+		Auths: []*hadoop_common.RpcSaslProto_SaslAuth{krbAuth},
+	}
+
+	if err = sendSaslMessage(client, con, &saslInitiateMessage); err != nil {
+		klog.Errorf("failed to send sasl initiate message: %v", err)
+		return err
+	}
+
+	saslResponseMessage, err := receiveSaslMessage(client, con)
+	if err != nil {
+		klog.Errorf("failed to receive sasl message: %v", err)
+		return err
+	}
+
+	// SASL CHALLENGE
+	if saslResponseMessage.GetState() == hadoop_common.RpcSaslProto_CHALLENGE {
+		// check hadoop qop
+		var serverWrapToken gssapi.WrapToken
+		if err := serverWrapToken.Unmarshal(saslResponseMessage.GetToken(), true); err != nil {
+			klog.Errorf("failed to unmarshal server challenge token: %v", err)
+			return err
+		}
```

Comment:
how about add verify?
```suggestion
		// Validate response
		isValid, err := serverWrapToken.Verify(krbAuth.encKey, keyusage.GSSAPI_ACCEPTOR_SEAL)
		if !isValid {
			return err
		}
```